proof/refine/ARM/SchedContextInv_R.thy

```diff
@@ -1220,17 +1220,409 @@ lemma refillUpdate_invs':
                          ko_wp_at'_def valid_refills_number'_def opt_map_red)
   done
 
-(* FIXME RT: preconditions can be reduced, this is what is available at the call site: *)
+lemma tcbSchedDequeue_valid_refills'[wp]:
+  "tcbSchedDequeue tcbPtr \<lbrace>valid_refills' scPtr\<rbrace>"
+  apply (clarsimp simp: tcbSchedDequeue_def)
+  apply (wpsimp wp: threadSet_wp threadGet_wp
+              simp: bitmap_fun_defs setQueue_def
+         | intro conjI impI)+
+  apply (fastforce simp: obj_at_simps valid_refills'_def opt_map_def split: option.splits)
+  done
+
+crunches tcbSchedDequeue, tcbReleaseRemove
+  for ksCurSc[wp]: "\<lambda>s. P (ksCurSc s)"
+  (wp: crunch_wps threadSet_wp simp: setQueue_def valid_refills'_def bitmap_fun_defs crunch_simps)
+
+lemma tcbReleaseRemove_valid_refills'[wp]:
+  "tcbReleaseRemove tcbPtr \<lbrace>valid_refills' scPtr\<rbrace>"
+  apply (clarsimp simp: tcbReleaseRemove_def)
+  apply (wpsimp wp: threadSet_wp threadGet_wp
+              simp: bitmap_fun_defs setReleaseQueue_def setReprogramTimer_def
+         | intro conjI impI)+
+    apply (fastforce simp: obj_at_simps valid_refills'_def opt_map_def split: option.splits)+
+  done
+
+lemma updateSchedContext_scBadge_update_active_sc_at'[wp]:
+  "updateSchedContext scPtr' (scBadge_update f) \<lbrace>active_sc_at' scPtr\<rbrace>"
+  apply (clarsimp simp: updateSchedContext_def)
+  apply (wpsimp wp: set_sc'.set_wp)
+  apply (fastforce simp: active_sc_at'_def obj_at_simps scBits_simps ps_clear_def)
+  done
+
+lemma updateSchedContext_scSporadic_update_active_sc_at'[wp]:
+  "updateSchedContext scPtr' (scSporadic_update f) \<lbrace>active_sc_at' scPtr\<rbrace>"
+  apply (clarsimp simp: updateSchedContext_def)
+  apply (wpsimp wp: set_sc'.set_wp)
+  apply (fastforce simp: active_sc_at'_def obj_at_simps scBits_simps ps_clear_def)
```

Comment:
We have `updateSchedContext_active_sc_at’` which should cover these cases.

Reply:
Removed